File: writepreview.c
#include "gitin.h"

#include <string.h>

static ssize_t writepandoc(FILE* fp, const struct repoinfo* info, const char* filename,
                           const char* content, ssize_t len, uint32_t contenthash,
                           const char* type) {

	ssize_t                 ret;
	struct callcached_param params = {
		.command     = pandoccmd,
		.cachename   = "preview",
		.content     = content,
		.ncontent    = len,
		.contenthash = contenthash,
		.fp          = fp,
		.info        = info,
		.nenviron    = 2,
		.environ     = (const char*[]){ "filename", filename, "type", type },
	};

	fprintf(fp, "<div id=\"preview\">\n");
	ret = callcached(&params);
	fprintf(fp, "</div>\n");

	return ret;
}

static ssize_t writetree(FILE* fp, const struct repoinfo* info, const char* content, ssize_t len,
                         uint32_t contenthash, const char* type) {

	ssize_t                 ret;
	struct callcached_param params = {
		.command     = configtreecmd,
		.cachename   = "preview",
		.content     = content,
		.ncontent    = len,
		.contenthash = contenthash,
		.fp          = fp,
		.info        = info,
		.nenviron    = 1,
		.environ     = (const char*[]){ "type", type },
	};

	fprintf(fp, "<div id=\"preview\">\n");
	ret = callcached(&params);
	fprintf(fp, "</div>\n");

	return ret;
}

static void writeimage(FILE* fp, int relpath, const char* filename) {
	fprintf(fp, "<div id=\"preview\">\n");
	hprintf(fp, "<img height=\"100px\" src=\"%rblob/%s\" />\n", relpath, filename);
	fprintf(fp, "</div>\n");
}


void writepreview(FILE* fp, const struct repoinfo* info, int relpath, const char* filename,
                  const char* content, ssize_t len, uint32_t contenthash) {
	char type[1024] = "", *param;

	for (int i = 0; filetypes[i][0] != NULL; i++) {
		if (endswith(filename, filetypes[i][0])) {
			strlcpy(type, filetypes[i][2], sizeof(type));
			break;
		}
	}

	if (!*type)
		return;

	if ((param = strchr(type, ':')))
		*param++ = '\0';

	if (!strcmp(type, "pandoc")) {
		if (!param && (param = strchr(filename, '.'))) {
			param++;
		}

		if (param)
			writepandoc(fp, info, filename, content, len, contenthash, param);
	} else if (!strcmp(type, "configtree")) {
		if (!param && (param = strchr(filename, '.'))) {
			param++;
		}

		if (param)
			writetree(fp, info, content, len, contenthash, param);
	} else if (!strcmp(type, "image")) {
		writeimage(fp, relpath, filename);
	}
}

File: writepreview.go
package gitin

import (
	"bytes"
	"encoding/xml"
	"errors"
	"fmt"
	"io"
	"path"
	"strings"
)

type Previewer = func(w io.Writer, blob *blobinfo, relpath int, param string) error

var ErrInvalidParam = errors.New("invalid parameter")

func writepandoc(fp io.Writer, blob *blobinfo, relpath int, typ string) error {
	if typ == "" {
		return fmt.Errorf("%w: %s", ErrInvalidParam, typ)
	}
	var params = executeinfo{
		command:     Config.Pandoccmd,
		cachename:   "preview",
		content:     blob.contents,
		contenthash: blob.hash,
		fp:          fp,
		environ:     map[string]string{"filename": blob.name, "type": typ},
	}

	fmt.Fprintf(fp, "<div class=\"preview\">\n")
	err := execute(&params)
	fmt.Fprintf(fp, "</div>\n")

	return err
}

func writepreviewtree(fp io.Writer, blob *blobinfo, relpath int, typ string) error {
	if typ == "" {
		return fmt.Errorf("%w: %s", ErrInvalidParam, typ)
	}
	fmt.Fprintf(fp, "<div class=\"preview\">\n")
	err := WriteConfigTree(fp, bytes.NewReader(blob.contents), typ)
	fmt.Fprintf(fp, "</div>\n")
	return err
}

func writeimage(fp io.Writer, blob *blobinfo, relpath int, filename string) error {
	fmt.Fprintf(fp, "<div class=\"preview\">\n")
	fmt.Fprintf(fp, "<img height=\"100px\" src=\"%sblob/%s\" />\n", strings.Repeat("../", relpath), filename)
	fmt.Fprintf(fp, "</div>\n")
	return nil
}

var Previewers = map[string]Previewer{
	"pandoc":     writepandoc,
	"configtree": writepreviewtree,
	"image":      writeimage,
}

type FileType struct {
	pattern string
	image   string
	preview string
}

func (ft FileType) Match(s string) bool {
	pre, suf, ok := strings.Cut(ft.pattern, "*")
	if !ok {
		return s == ft.pattern
	}
	return strings.HasPrefix(s, pre) && strings.HasSuffix(s, suf)
}

func writepreview(fp io.Writer, relpath int, blob *blobinfo) error {
	var typ string
	for _, ft := range filetypes {
		if ft.Match(blob.name) {
			typ = ft.preview
			break
		}
	}

	typ, param, _ := strings.Cut(typ, ":")
	if param == "" {
		param = strings.TrimLeft(path.Ext(blob.name), ".")
	}

	if fun, ok := Previewers[typ]; ok {
		err := fun(fp, blob, relpath, param)
		/* either nil or unknown error */
		if !errors.Is(err, ErrInvalidParam) {
			return nil
		}
	}

	fmt.Fprintf(fp, "<div class=\"preview\">\n<pre>\n")
	err := xml.EscapeText(fp, blob.contents)
	fmt.Fprintf(fp, "</pre>\n</div>\n")
	return err
}
